### ISSUE TITLE
Reading SPECT DICOM raw data

### ISSUE THREAD

**Issue by lukepolson** [UCL/STIR#1089]
Hi all,

I'm trying to open projection data stored in a DICOM file. I've established a conda environment using:

https://stir.sourceforge.net/wiki/index.php/Installing_STIR_with_conda

When I try to open the projection data from the DICOM file using STIR in python, I get an error:
`projdata = stir.ProjData.read_from_file('test.dcm')`

`
ERROR: 
ProjData::read_from_file could not read projection data test.dcm.
Unsupported file format? Aborting.
`

I see in the user manual that if one compiles STIR so that it can "see" the ITK library, then one is able to open DICOM files. Is this conda environment compiled so that it can see ITK, and if not, is there another image which is?

In summary: how can I open a DICOM file of projection data using the python STIR package?

EDIT: Running 

`stir.FloatVoxelsOnCartesianGrid_read_from_file('test.dcm')`

works, but the data is not on a cartesian grid, since it represents projection data. Perhaps there's some fields missing, or a way to convert from `FloatVoxelsOnCartesianGrid` to `ProjData`?

**Comment by KrisThielemans** [comment]
Hi Luke

at present, ITK is only used to read images. The conda version of STIR should therefore be able to read DICOM images, but sadly not SPECT DICOM acquisition data.

@danieldeidda you have this functionality somewhere. Is this something that we can share?

**Comment by danieldeidda** [comment]
Hi, 
I do have this functionality but as part of  a private repo. I can however, make a new STIR utility and a PR

**Comment by danieldeidda** [comment]
The functionality is now here #1182 feel free to try it on

**Comment by KrisThielemans** [comment]
#1282 is now merged. Feedback welcome

**Comment by smanwell** [comment]
I am currently trying to make use of the SPECT_dicom_to_interfile utilitiy with a projection dataset where there are two detector heads that are separated by 180 degrees with respect to one another and I have noticed a few (what I think are...) issues.

he SPECT_dicom_to_interfile tool looks at the DICOM Attribute "Number of Frames in Rotation" (0x0054, 0x0053) to determine the number of projections contains in the dataset. For this particular dataset that value is 60. For this dataset, however, there are two items within its Detector Information Sequence corresponding the projection images acquired by its two detector heads. Should the total number of projections should not be 120 (i.e. 60 per detector head)?

This dataset also contains the projection data for lower and upper scatter windows. Running this utlility as is, triggers the following warning message

`stir::warning(boost::format("num_projections (%1%) * num_energy_windows (%2%) != num_frames (%3%)\n"
                            "This could be gated or dynamic data. The interfile header(s) will be incorrect!")
              % spect.num_of_projections % spect.num_energy_windows % spect.num_frames);`

as the total number of frames is 360, the number of energy windows is 3 and the number of projections was taken as 60.

I tried modifying the function open_dicom_file( bool is_planar ) to instead calculate the number of projections as the product of the "Number of Frames in Rotation" and the "Number of Detectors" to workaround the warning and to ensure the correct pixel buffer size is used when generating header files for each energy window. Which seems to work ok until I tried using the photopeak interfile in a reconstruction.

The reconstruction throws the warning: "WARNING: Interfile error: number of projections must be consistent with radius vector length"

Looking at the interfile header for the photopeak energy window I see that the "radii" field contains only the Radial Positions for a single detector head.  Looking again at the function open_dicom_file( bool is_planar ) I see that the call to

`GetDICOMTagInfo( file, gdcm::Tag(0x0018,0x1142), radius_as_string )`

is made only once with the default value for its sequence_idx argument (1). Presumably, the retrieval of the radial positions should be called within a loop over the number of detectors (or equivalently over the cardinality of the Detector Information Sequence).

I'll try making this amendment myself to see if the reconstruction proceeds successfully or not, but I am curious to hear thoughts about this from the STIR team.

Thanks in advance!

**Comment by KrisThielemans** [comment]
@smanwell that seems correct. This is mentioned in #1184 but you have added more detail here. Possibly best to continue there, although that doesn't matter too much.
